src/coord/axisTickLabelBuilder.ts

```diff
@@ -191,95 +233,309 @@ function makeRealNumberLabels(axis: Axis) {
+}
+
+function calculateLabelUnionRect(
+    axis: Axis,
+    getLabel: (i: number) => string,
+    tickCount: number,
+    step: number,
+    rotation: number
+) {
+    const labelModel = axis.getLabelModel();
+    const padding = getOptionLabelPadding(labelModel);
+    const isHorizontal = isHorizontalAxis(axis);
+    let rect;
+    for (let i = 0; i < tickCount; i += step) {
+        const labelRect = rotateLabelRect(labelModel.getTextRect(getLabel(i)), rotation, padding, isHorizontal);
+        rect ? rect.union(labelRect) : (rect = labelRect);
     }
-    let tickValue = ordinalExtent[0];
+
+    return rect;
+}
+
+function rotateLabelRect(
+    originalRect: RectLike,
+    rotation: number,
+    padding: NormedPadding,
+    isHorizontal: boolean
+) {
+    const { bounds, offset } = rotateLabel(originalRect, rotation);
+    const { width, height } = applyPadding(bounds, padding);
+    return new BoundingRect(
+        0, 0,
+        width - (isHorizontal ? 0 : offset.x),
+        height - (isHorizontal ? offset.y : 0)
+    );
+}
+
+function makeAutoCategoryLayout(axis: Axis, interval?: number) {
+    const result = inner(axis).autoLayout;
+    return result != null && (interval === undefined || result.interval === interval)
+        ? result
+        : (inner(axis).autoLayout = axis.calculateCategoryAutoLayout(interval));
+}
+
+function calculateUnitDimensions(axis: Axis): Size {
+    const rotation = getAxisRotate(axis) * RADIAN;
+
+    const ordinalScale = axis.scale as OrdinalScale;
+    const ordinalExtent = ordinalScale.getExtent();
+    const tickValue = ordinalExtent[0];
+
     const unitSpan = axis.dataToCoord(tickValue + 1) - axis.dataToCoord(tickValue);
-    const unitW = Math.abs(unitSpan * Math.cos(rotation));
-    const unitH = Math.abs(unitSpan * Math.sin(rotation));
+    return {
+        width: Math.abs(unitSpan * Math.cos(rotation)),
+        height: Math.abs(unitSpan * Math.sin(rotation))
+    };
+}
+
+function calculateMaxLabelDimensions(axis: Axis): Size {
+    const labelFormatter = makeLabelFormatter(axis);
+    const axisLabelModel = axis.getLabelModel();
+
+    const ordinalScale = axis.scale as OrdinalScale;
+    const ordinalExtent = ordinalScale.getExtent();
+    const step = layoutScaleStep(ordinalScale.count(), axisLabelModel);
 
     let maxW = 0;
     let maxH = 0;
 
     // Caution: Performance sensitive for large category data.
     // Consider dataZoom, we should make appropriate step to avoid O(n) loop.
-    for (; tickValue <= ordinalExtent[1]; tickValue += step) {
-        let width = 0;
-        let height = 0;
-
-        // Not precise, do not consider align and vertical align
-        // and each distance from axis line yet.
-        const rect = textContain.getBoundingRect(
-            labelFormatter({ value: tickValue }), params.font, 'center', 'top'
-        );
-        // Magic number
-        width = rect.width * 1.3;
-        height = rect.height * 1.3;
```

Comment:
The magic 1.3 coefficient here was effectively a workaround for a subtle bug that caused an incorrect axis pixel extent to be used when calculating the category axis' unit dimensions. Please see below for the bug fix.